### ISSUE TITLE
[Regression][SM100/SM103] TRTLLM NVFP4 MoE emits NaNs in v0.6.16.post4/v0.6.17 (v0.6.16rc4 passes)

### ISSUE THREAD

**Issue by zianglih** [flashinfer-ai/flashinfer#4486]
## Summary

TRTLLM-Gen per-token NVFP4 MoE produces NaNs on Blackwell with FlashInfer `0.6.16.post4` and `0.6.17`, while the same SGLang checkout and workload pass with `0.6.15.post1` and `0.6.16rc4`.

The measured version boundary is **v0.6.16rc4 PASS -> v0.6.16.post4 FAIL**. I did not directly run rc5. Source/tag containment makes **v0.6.16rc5 the inferred first affected release**: [#4252 / `c4053a28`](https://github.com/flashinfer-ai/flashinfer/pull/4252) first appears in rc5 and changes the pinned TRTLLM BMM artifact as well as the advertised FP4 tile set. The same underlying change is [#4180 / `a34a735b`](https://github.com/flashinfer-ai/flashinfer/pull/4180) on the v0.6.17 line.

## Environment

```text
Hardware: 8x NVIDIA B300 SXM6 AC (SM103, 267.7 GiB, 148 SMs); server uses 4 GPUs (TP4/EP4)
Driver: 590.48.01
Container: lmsysorg/sglang:nightly-dev-cu13-20260812-c7c03ec5
amd64 digest: sha256:e948f2d4adb7d3ca8336d855f667ab26ab27bc7515702336e3c6f7093121bf23
OS: Ubuntu 24.04.4 LTS, Linux 6.8.0-137-generic, glibc 2.39
Python: 3.12.3
PyTorch: 2.13.0+cu130
CUDA: 13.0 (nvcc 13.0.88)
NCCL: 2.29.7
cuDNN frontend/backend: 1.27.0 / 9.20.0
FlashInfer 0.6.17 commit: a0a6b019b9b27d49d209f85d028a1ae5a9b347d7
FlashInfer packages: flashinfer-python==0.6.17, flashinfer-cubin==0.6.17,
                     flashinfer-jit-cache==0.6.17+cu130
SGLang: 3f3002d4a88d964f2f19ef501755f994cd5d25cc
Model: Qwen/Qwen3-30B-A3B-Instruct-2507-FP8
Backend / quantization: flashinfer_trtllm / nvfp4_online
```

The full output of `python3 -m flashinfer.collect_env` is available on request. The container image declares stale `FLASHINFER_VERSION=0.6.15.post1`; package metadata, `flashinfer.__version__`, `flashinfer show-config`, and `collect_env` all confirmed the aligned runtime packages shown above.

## Reproduction

The model requires Hugging Face authentication. Install SGLang first, then replace all three FlashInfer packages together. Repeat with `VERSION=0.6.16rc4`, `0.6.16.post4`, and `0.6.17` using a fresh server process and cache directory.

```bash
git clone https://github.com/zianglih/sglang.git
cd sglang
git checkout 3f3002d4a88d964f2f19ef501755f994cd5d25cc
python3 -m pip install -e python

VERSION=0.6.17
python3 -m pip uninstall -y \
  flashinfer-python flashinfer-cubin flashinfer-jit-cache
rm -rf /root/.cache/flashinfer
python3 -m pip install --no-cache-dir --no-deps \
  "flashinfer-python==${VERSION}"
python3 -m pip install --no-cache-dir --no-deps \
  --index-url https://flashinfer.ai/whl \
  "flashinfer-cubin==${VERSION}"
python3 -m pip install --no-cache-dir --no-deps \
  --index-url https://flashinfer.ai/whl/cu130 \
  "flashinfer-jit-cache==${VERSION}+cu130"
python3 -m pip check
python3 -m flashinfer show-config

cd test
SGLANG_ENABLE_ASYNC_ASSERT=true \
python3 registered/backends/test_flashinfer_nvfp4_online_moe_backend.py \
  TestFlashinferTrtllmGenMoeBackendNvFp4Online.test_gsm8k -f
```

The registered test launches:

```text
sglang serve \
  --model-path Qwen/Qwen3-30B-A3B-Instruct-2507-FP8 \
  --moe-runner-backend flashinfer_trtllm \
  --cuda-graph-max-bs-decode 128 \
  --tp-size 4 --ep-size 4 \
  --quantization nvfp4_online \
  --mem-fraction-static 0.7
```

with:

```text
FLASHINFER_NVFP4_4OVER6=1
FLASHINFER_NVFP4_4OVER6_ERR_MODE=MSE
FLASHINFER_NVFP4_4OVER6_ERR_USE_FAST_MATH=1
FLASHINFER_NVFP4_4OVER6_E4M3_USE_256=1
SGLANG_FP4_IGNORED_LAYERS=shared_expert,model.layers.40,model.layers.41,model.layers.42,model.layers.43,model.layers.44,model.layers.45,model.layers.46,model.layers.47
SGLANG_ENABLE_JIT_DEEPGEMM=False
SGLANG_ENABLE_ASYNC_ASSERT=true
```

The evaluation is 200 GSM8K examples, 128 client threads, completion API, `max_tokens=512`. SGLang automatically logs `--disable-shared-experts-fusion is automatically set`; shared-expert fusion is not active.

## Confirmed package-only bisect

The SGLang SHA, image, model, arguments, environment, hardware, and full workload were held constant:

| Aligned FlashInfer trio | Result | GSM8K score | Latency | Throughput | First observed failure |
|---|---:|---:|---:|---:|---|
| `0.6.15.post1` | PASS | `0.960` | `4.2688 s` | `6747.30 tok/s` | none |
| `0.6.16rc4` | PASS | `0.945` | `4.7273 s` | `6153.64 tok/s` | none |
| `0.6.16.post4` | FAIL | `0.000` | `147.96 s` | n/a | pure prefill, `M=1792`; NaNs on all four ranks |
| `0.6.17` | FAIL | `0.000` | n/a | n/a | first concurrent prefill; reduced trace reaches `M=896`; NaNs on all four ranks |

This brackets the measured runtime regression after `0.6.16rc4`. Source/tag inspection points to rc5 as the first line carrying the corresponding artifact/launcher update, but rc5 itself has not yet been executed.

## Independent B200 reproduction

The same registered test independently reproduces the first-`M=896`, all-rank NaN symptom on four NVIDIA B200 GPUs (SM100):

- SGLang: `8531fe0a8ca4e3b6d92b4134be399aaac261c636`
- Driver: `580.126.09`
- PyTorch: `2.13.0+cu130`
- Aligned FlashInfer trio: `0.6.17` / `0.6.17` / `0.6.17+cu130`
- Job: https://github.com/sgl-project/sglang/actions/runs/31607763933/job/94151038518

The first `M=896` prefill logs `Assertion 'NaN detected! sampler: next_token_logits' failed.` on all four ranks and GSM8K score `0.000`. The later CUDA abort/coredumps are downstream effects and are not separately root-caused here.

## Control matrix on 0.6.17 / B300

| Single change | Scope | Result |
|---|---|---|
| default | 1 example, 1 thread, `max_tokens=512` | PASS |
| default | full 200/128/512 | FAIL, score `0.000` |
| `--disable-prefill-cuda-graph` | full 200/128/512 | FAIL |
| Triton attention + page size 64 | full 200/128/512 | FAIL |
| `flashinfer_trtllm_routed` | reduced concurrent run | FAIL |
| deferred finalize disabled | reduced concurrent run | FAIL |
| `SGLANG_TRTLLM_MOE_PDL_MAX_TOKENS=0` | reduced concurrent run | FAIL |
| `--enforce-disable-flashinfer-allreduce-fusion` | full 200/128/512 | FAIL |
| `SGLANG_FORCE_COARSE_WAR_BARRIER=1` | full 200/128/512 | FAIL |
| `--disable-flashinfer-autotune` | full 200/128/512 | FAIL |
| force exact `[128,19]` through `M=16384`, v0.6.17 cubins | full 200/128/512, prefill graph disabled | PASS: score `0.955`, no non-finite marker or async assertion |
| same `[128,19]`, substitute the exact same-name v0.6.15 FC1+FC2 cubins | 24 requests / 24 threads / `max_tokens=1`, prefill graph disabled | PASS |
| force `[192,1]`, v0.6.17 cubins | startup autotune, `M=4096` | FAIL: non-finite finalized MoE output on all ranks; serving later hits async device assertions |
| force `[192,3]`, v0.6.17 cubins | startup autotune, `M=4096` | FAIL identically |
| device sync at FP4 entry; stream sync after routing, FC1, explicit quantization, or FC2 | reduced eager run, one probe per run | FAIL at every boundary |
| `--disable-overlap-schedule` | full 200/128/512 | PASS, score `0.955` |
| `CUDA_LAUNCH_BLOCKING=1` | full 200/128/512 | PASS, score `0.940` |

Attention and prefill CUDA graph capture are not required. PDL, routed-vs-FromLogits dispatch, deferred finalize, FlashInfer all-reduce fusion, and SGLang's coarse WAR barrier are insufficient explanations. The forced tile-192 tactics become non-finite during FlashInfer's startup autotune, before serving requests, so SGLang request concurrency/overlap is not required to reproduce the kernel defect. Serialization can still mask the default serving failure.

## Artifact and tactic evidence

| Release line | `TRTLLM_GEN_BMM` | Manifest checksum |
|---|---|---|
| `0.6.15.post1`, `0.6.16rc4` | `b368d003.../batched_gemm-da58956-b4ac80e/` | `d0178cd4...` |
| `0.6.16rc5`, `0.6.16.post4`, `0.6.17` | `5988e15c.../batched_gemm-3d40263-3e19f0a/` | `b19ed6c8...` |

The old manifest has 2,516 configs and the new one has 3,418. All 2,516 old function names remain, but every corresponding cubin SHA changes; 902 configs are added. The same change adds FP4 `tile_N=192` to fallback enumeration.

The first-nonfinite probe narrows the runtime boundary further. On the failing concurrent run, layer 0 receives finite BF16 input, block scales, per-token scales, and router logits, while the finalized BF16 tensor returned from `flashinfer::trtllm_fp4_block_scale_moe` is non-finite on all four ranks. The corrupt MoE invocation is the aggregated batch (`M=15605` in that trace), not the nearby scheduler line that reports 896 new tokens.

Capture-aware synchronization probes also ran on every rank: `cudaDeviceSynchronize()` at FP4 FFI entry and `cudaStreamSynchronize()` after routing, FC1, explicit BF16-to-NVFP4 quantization, and FC2. Every variant still returned non-finite output. This rules out a simple missing completion edge between those high-level stages or unfinished work already enqueued before the call; it does not identify corruption inside a compound BMM stage.

The v0.6.17 tile-192 tactics are the Cartesian product below:

| Tactic | FC1 raw config | FC2 raw config | Shape-valid for `I=768` | Result |
|---|---:|---:|---:|---|
| `[192,0]` | 521 (`u2`, 2-CTA) | 779 (`u2`, 1-CTA) | no | rejected as an invalid tactic |
| `[192,1]` | 521 (`u2`, 2-CTA) | 787 (non-`u2`, 1-CTA) | yes | non-finite |
| `[192,2]` | 529 (non-`u2`, 2-CTA) | 779 (`u2`, 1-CTA) | no | not runnable for this shape |
| `[192,3]` | 529 (non-`u2`, 2-CTA) | 787 (non-`u2`, 1-CTA) | yes | non-finite |

Both shape-valid tile-192 tactics fail during the real `M=4096` dummy-model forward used by startup autotuning. They use different FC1 cubins but share FC2 raw config 787. In contrast, forced tile-128 `[128,19]` stays finite and passes the full 200-example test with score `0.955` using the v0.6.17 cubins; substituting its exact same-name v0.6.15 FC1+FC2 pair also stays finite in the reduced crash reproducer. The old artifact has no tile-192 cubins to substitute directly.

## Root-cause status

Confirmed:

- This is a FlashInfer-version-dependent regression: package-only replacement flips pass/fail on identical SGLang code.
- Layer 0 enters FlashInfer FP4 MoE with finite inputs/scales/router data and returns non-finite BF16 output.
- The release/source boundary and available runtime tactics both change at the #4252/#4180 artifact update.
- Both shape-valid tile-192 tactics fail before serving starts; tile-128 `[128,19]` passes the full correctness test.

Strong working hypothesis, not yet a final root cause:

- a correctness defect in the newly exposed tile-192 TRTLLM_GEN BMM path;
- FC2 raw config 787 or its tile-192 integration contract is the strongest common-variable localization, because both valid tactics share that FC2 while changing FC1. A finalized-output probe does not yet prove FC2 is the first internal tensor to become non-finite.

Relevant prior context, not the regression's introduction:

- [#3971](https://github.com/flashinfer-ai/flashinfer/issues/3971) reports an artifact-only TRTLLM BMM regression where disabling PDL did not help and substituting the old artifact did.
- [#3973](https://github.com/flashinfer-ai/flashinfer/pull/3973) refreshed cubins to fix a 2-CTA hang before the passing `0.6.15.post1` artifact.
- [#3991](https://github.com/flashinfer-ai/flashinfer/pull/3991) is an older, still-open 2-CTA filtering proposal. Both tile-192 FC1 choices are 2-CTA, so filtering those kernels plus suppressing the now-empty tile-192 family would avoid this path, but it does not identify the generator-side defect.

## Current upstream status

As of current main [`2d132e68`](https://github.com/flashinfer-ai/flashinfer/commit/2d132e68c9c70e17efa2a0c81c56508ba2cf421d), the shipped pin remains `5988e15c.../batched_gemm-3d40263-3e19f0a/`, and NVFP4 tile 192 remains advertised. I found no merged post-0.6.17 change targeting this path.

Open [#4281](https://github.com/flashinfer-ai/flashinfer/pull/4281) (head `e85337e0`) refactors BMM dispatch and repins to `d2659297.../batched_gemm-4868e72-cacd555`. It is not in main and has not been tested with this reproducer. Its exact tile-192 FC1/FC2 function names remain but both cubins are rebuilt, so it is a valuable A/B, not an evidenced fix.

## Expected behavior

TRTLLM NVFP4 MoE outputs should remain finite and the unchanged end-to-end test should retain its `>0.90` GSM8K score across the dependency update.

## Requested help

1. Is tile-192 FC2 raw config 787 known to have an SM100/SM103 per-token-NVFP4 correctness issue?
2. Is the new artifact in #4281 intended to address this path?
3. Can maintainers provide a fixed artifact or a supported way to suppress tile 192 so fallback selects a valid tile-128/256 tactic?

Verified mitigations are pinning the aligned trio to `0.6.15.post1` / tested `0.6.16rc4`, forcing valid tile-128 tactic `[128,19]` (full score `0.955`), or using SGLang `--disable-overlap-schedule`. `CUDA_LAUNCH_BLOCKING=1` is diagnostic only.

Linked SGLang integration issue: https://github.com/sgl-project/sglang/issues/34629

No exact duplicate was found for this version/shape/NaN regression. #3971/#3973/#3991 are the closest artifact/2-CTA context but report or address an older hang.

**Comment by zianglih** [comment]
### Root-cause correction and completed tactic A/B

I corrected the issue body to withdraw two earlier inferences: the corrected cache shows that forced tile 128 passes, and SGLang request overlap is not required to produce the non-finite result.

- At layer 0, BF16 input, block scales, per-token scales, and router logits are finite immediately before `flashinfer::trtllm_fp4_block_scale_moe`; its finalized BF16 output is non-finite on all ranks.
- Entry/device and routing/FC1/quant/FC2 stream fences all still fail, excluding a simple missing high-level completion edge.
- With v0.6.17 otherwise unchanged, forced `[128,19]` through the 16K profile passes the complete 200-example/128-thread test: score `0.955`, latency `11.087 s`, output throughput `2580.322 tok/s`, no non-finite marker, and no async assertion.
- Both shape-valid tile-192 tactics fail during the real `M=4096` dummy-model forward used by startup autotuning:
  - `[192,1]` = FC1 raw 521 + FC2 raw 787
  - `[192,3]` = FC1 raw 529 + FC2 raw 787
- `[192,0]`/`[192,2]` use FC2 raw 779 and are invalid for this `I=768` shape, so raw 787 is the only runnable tile-192 FC2.

This localizes the regression to the newly exposed tile-192 TRTLLM_GEN BMM path. FC2 raw 787 or its tile-192 integration contract is the strongest common-variable lead, although the finalized-output probe does not yet prove FC2 is the first internal tensor to become non-finite.

The validated short-term mitigation is to suppress tile 192 / force a valid tile-128 tactic, or pin the aligned FlashInfer trio to `0.6.16rc4` or earlier. Existing autotune caches must be cleared or regenerated after filtering the tactic.

**Comment by leejnau** [comment]
I believe this PR is addressing it:  https://github.com/flashinfer-ai/flashinfer/pull/4441

**Comment by zianglih** [comment]
@leejnau no this is a regression in trtllm-gen per-token fp4 moe

**Comment by b8zhong** [comment]
@leejnau Yes. It's an bug that needs to be fixed (probably from cubin side)

**Comment by bkryu** [comment]
cc @aleozlx

**Comment by aleozlx** [comment]
adding @cindyzxq for checking if we cover this testing pre-release
